docs/docs/guides/smart_contracts/testing_contracts/index.md

```diff
@@ -0,0 +1,180 @@
+
+## TXE
+
+In order to interact with the protocol, Aztec contracts leverage the power of oracles: functions that reach out to the outside world and are able to query and manipulate data outside of itself. The values returned by oracles are then constrained inside Noir and modifications to the blockchain state are later verified to adhere to the protocol rules by our kernel circuits.
+
+However, all of this is often not necessary to ensure the contract logic itself is sound. All that we need is an entity to provide values consistent with real execution. This is where our TXE (Testing eXecution Environment) comes in!
+
+TXE is a JSON RPC server much like PXE, but provides an extra set of oracle functions called `cheatcodes` that allow developers to manipulate the state of the chain and simulate contract execution. Since TXE skips most of the checks, block building and other intricacies of the Aztec protocol, it is much faster to run than simulating everything in the sandbox.
+
+### Running TXE
+
+In order to use the TXE, it must be running on a known address. Assuming the default `http://localhost:8080`, contract tests would be run with:
+
+`nargo test --oracle-resolver http://localhost:8080`
+
+:::warning
+Since TXE tests are written in Noir and executed with `nargo`, they all run in parallel. This also means every test creates their own isolated environment, so state modifications are local to each one of them.
+
+Executing many tests in parallel might slow processing of the RPC calls down to the point of making them timeout. To control this timeout the `NARGO_FOREIGN_CALL_TIMEOUT` env variable is used.
+:::
+
+### Writing TXE contract tests in Noir
+
+`aztec-nr` provides an utility class called `TestEnvironment`, that should take care of the most common operations needed to setup contract testing. Setting up a new test environment with `TestEnvironment::new()` **will reset the current test's TXE state**.
+
+You can find all of the methods available in the `TestEnvironment` [here](https://github.com/AztecProtocol/aztec-packages/blob/#include_aztec_version/noir-projects/aztec-nr/aztec/src/test/helpers/test_environment.nr).
+
+#include_code txe_test_increment /noir-projects/noir-contracts/contracts/counter_contract/src/main.nr rust
+
+:::warning
+Tests run significantly faster as `unconstrained` functions. This means we generate bytecode (Brillig) and not circuits (ACIR), which _should_ yield exactly the same results. Any other behavior is considered a bug.
+:::
+
+#### Imports
+
+Writing tests in contracts requires importing additional modules. Here are the modules that are needed for testing the increment function in the counter contract.
+
+#include_code test_imports /noir-projects/noir-contracts/contracts/counter_contract/src/main.nr rust
+
+#### Deploying contracts
+
+```rust
+let deployer = env.deploy("path_to_contract_ts_interface");
```

Comment:
@Thunkar it looks like `env.deploy` is always called with and initializer method like `env.deploy("@aztec/noir-contracts.js/Child").without_initializer();`. I'll update this section to follow a similar syntax and a note unless you have any objections.

Reply:
Whoops, is the part underneath that's wrong, it should read `deployer.with_private_initializer...`, `deployer.with_public_initializer...`

I separated it to make clear all of those are possible, but used `env` instead of `deployer` 🤦